<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2]:
import pandas as pd
import psycopg2

In [ ]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [6]:
# текст запроса
""" Считаем количество вакансий в sql vacancies"""

query_3_1 = f'''select
            count(id) 
            from vacancies
            
'''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_22252\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [8]:
# текст запроса
""" Считаем количество работодалей в sql employers"""
query_3_2 = f'''select
            count(id)
            from employers
            
'''

In [31]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [32]:
# текст запроса
""" Считаем количество регионов в sql areas"""
query_3_3 = f'''select
            count(id)
            from areas
            
'''

In [34]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [35]:
# текст запроса
""" Считаем количество сфер детельности в sql industries"""

query_3_4 = f'''select
            count(id)
            from industries
            
'''

In [36]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

In [25]:
# выводы по предварительному анализу данных

""" _summary_
Соединяем две таблица, чтобы посмотреть ТОП 5 работодателй с максимальным количеством, размещенных вакансий 

"""


query_3_5 = f'''select
            count(v.id),
            e.name
            from vacancies v
            join employers e on v.employer_id =e.id
            group by e.name
            order by 1 desc
            limit 5
            
'''


In [26]:
df = pd.read_sql_query(query_3_5, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\2214534321.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_5, connection)


count           name
0   1933         Яндекс
1    491     Ростелеком
2    444       Тинькофф
3    428           СБЕР
4    331  Газпром нефть

In [23]:
""" _summary_
Соединяем две таблица, чтобы посмотреть ТОП 5 регионов с максимальным количеством, размещенных вакансий 

"""


query_3_6 = f'''select
            count(v.id),
            a.name
            from vacancies v
            join areas a on v.area_id =a.id
            group by a.name
            order by 1 desc
            limit 5
            
'''

In [24]:
df = pd.read_sql_query(query_3_6, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\4276945241.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_6, connection)


count             name
0   5333           Москва
1   2851  Санкт-Петербург
2   2112            Минск
3   2006      Новосибирск
4   1892           Алматы

In [27]:
""" _summary_
Выбираем ТОП 5 самых популярных вакансий 

"""


query_3_7 = f'''select
            count(v.id),
            v.name
            from vacancies v
            group by v.name
            order by 1 desc
            limit 5
            
'''

In [28]:
df = pd.read_sql_query(query_3_7, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\1176558524.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_7, connection)


count                              name
0   1761           Системный администратор
1   1153                    Программист 1С
2    488                   Бизнес-аналитик
3    457  Специалист технической поддержки
4    432                Системный аналитик

Предварительный анализ данных показал, что в **1362** регионах **23 501** работодателя из **294** сфер деятельности опубликовали **49 197** вакансий.

Из которых:
- Топ 5 регионов: Москва (5333), Санкт-Петербург (2851), Минск (2112), Новосибирск (2006), Алматы (1892).
- Топ 5 компаний: Яндекс (1933), Ростелеком (491), Тинькофф (444), СБЕР (428), Газпром нефть (331).
- Топ 5 профессий: Системный администратор (1761), Програмист 1С (1153), Бизнес-аналитик (488), Специалист тех.поддержки (457), Системный аналитик (432).

По предварительному анализу данных можно сделать следующие выводы:
**
 


# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [43]:
# текст запроса
""" _summary_
Соединяем две таблицы по id, чтобы вывести количество вакансий в регионах.
Группируем запрос по id регионов (a.id).
Сортируем количество вакансий (count(v.id)) по убыванию 

"""

query_4_1 = f'''select
            count(v.id),
            a.name
            from vacancies v
            join areas a on v.area_id = a.id
            group by a.id
            order by 1 desc
            
'''

In [44]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


count                      name
0     5333                    Москва
1     2851           Санкт-Петербург
2     2112                     Минск
3     2006               Новосибирск
4     1892                    Алматы
..     ...                       ...
764      1                    Кизляр
765      1                    Джизак
766      1                    Эртиль
767      1                  Арсеньев
768      1  Тура (Красноярский край)

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [76]:
# текст запроса

""" _summary_
Выводим количество вакансий при условии, что нижний (salary_from) и верхний порог зарплат (salary_to) не NULL 

"""
query_4_2 = f'''select
            count(id)
            from vacancies
            where ((salary_from is not NULL) OR (salary_to is not NULL))
            
            
'''

In [77]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [84]:
# текст запроса

""" _summary_
Считаем средние значения нижнего (salary_from) и верхнего(salary_to)  порогов зарплат 

"""
query_4_3 = f'''select
            AVG(salary_from),
            AVG(salary_to)
            from vacancies
            
            
            
'''

In [85]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg            avg
0  71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [88]:
# текст запроса

""" _summary_
Выводим количество вакансии для разных вариантов графика и занятости.
Группируем данные по графику и занятости.
Сортируем по количеству вакансий по убыванию.
"""
query_4_4 = f'''select
            schedule,
            employment,
            count(id)
            from vacancies
            Group by schedule, employment
            order by 3  desc
            
            
'''

In [89]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [33]:
# текст запроса
""" _summary_
Считаем количество вакансий по каждому виду требуемого опыта работы.
Группируем данные по видам требуемого опыта.
Сортируем данные по количеству вакансий по возрастанию

"""

query_4_5 = f'''select
            experience,
            count(id)
            from vacancies
            group by experience
            order by count(id) asc
            
            
'''

In [265]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

In [ ]:
# выводы по детальному анализу вакансий


- Большинство вакансий предоставлены от московский работодателей (5333).

- Работодатели в основном ищут специалистов на полный рбочий день с полной занятостью (35367), на удаленную работу с полной занятостью ищут в 5 раз меньше соискателей (7082). Скорее всего это связано с непопулярностью работы удаленно и особенностью вакансий. Так системный администратор (самая популярная вакансия) не всегда может выполнять свои обязанности удаленно. Еще может быть весомым тот фактор, что работников легче контролировать в офисе. 

- Высокий показатель вакансий с требуемым опытом работы от 1 до 3 лет (26152) - скорее всего показатель того, что работодатели готовы рассматривать специалистов с минимальным опытом. Такие специалисты еще не требуют высоких ЗП, и готовы развиваться в профессии. А вот специалисты с опытом от 6 лет(1337) менее популярны - как раз потому что они уже знают многое, скорее всего требуют больших ЗП и, возможно, им сложно перестраиваться под новых работодателей.

- Средние показатели зарплатной вилки: нижний порог - около 71 тыс., верхний - около 110 тыс. Опять же подтверждает, что работодатели ищут специалистов от 1 до 3 лет, которые не имеют высоких зарплатных ожиданий

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [263]:
# текст запроса

""" _summary_
Выводим данные по названиям работодателей, разместивших вакансии.
Объединяем employers и vacancies.
Группируем по названию работодателя.
Сортируем по количеству вакансий по убыванию.

"""
query_5_1 = f'''select
            e.name
            from employers e
            join vacancies v on v.employer_id = e.id
            group by e.name
            order by count(v.id) desc
            limit 5
            
            
'''

In [266]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name
0         Яндекс
1     Ростелеком
2       Тинькофф
3           СБЕР
4  Газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [6]:
# текст запроса

""" _summary_
Выводим данные по названиям работодателей, разместивших вакансии.
Объединяем employers и vacancies.
Группируем по названию работодателя.
Сортируем по количеству вакансий по убыванию.
"""
 
query_5_2 = f'''select
            a.name,
                count(distinct e.id) as employers_cnt,
                count(distinct v.id) as vacancies_cnt
            from areas a
                left join employers e on a.id = e.area 
                left join vacancies v on a.id = v.area_id
            where v.area_id is null
            group by a.name
            order by 2 desc
            
            
'''

In [7]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  employers_cnt  vacancies_cnt
0                            Россия            410              0
1                         Казахстан            207              0
2                Московская область             75              0
3                Краснодарский край             19              0
4                          Беларусь             18              0
..                              ...            ...            ...
588                        Кувандык              0              0
589                        Куйбышев              0              0
590  Куйбышево (Ростовская область)              0              0
591                        Култаево              0              0
592                        Кульсары              0              0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [60]:
# текст запроса

""" _summary_
Выводим данные по количеству регионов count(distinct v.area_id), в которых работодатели публикют вакансии.
Соединяем таблицы employers и vacancies по id работодателя.
Группируем данные по названию работодателей.
Сортируем количество регионов по убыванию.
"""

query_5_3 = f'''select
            e.name,
            count(distinct v.area_id)
            from employers e
            join vacancies v on e.id = employer_id  
            group by e.name
            order by 2 desc
            
            
'''

In [61]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
14761                  UniSol      1
14762            UNISTORY LLC      1
14763                   UNIT6      1
14764     United Distribution      1
14765              UNITEL LLC      1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [66]:
# текст запроса
""" _summary_
Выводим количество работодателей (count(e.name)), у которых отсутствует сфера деятельности (s.employer_id is NULL).
Соединяем таблицы employers и employers_industries по id работодателя.

"""

query_5_4 = f'''select
            count(e.name)
            from employers e
            left join employers_industries s on e.id=s.employer_id
            where s.employer_id is NULL

            
            
'''

In [67]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [72]:
# текст запроса
""" _summary_
Выводим название компании (e.name), у которой указано 4 сферы деятельности (count(i.name) = 4).
Соединяем таблицы employers, employers_industries и  industries по id работодателя, и по id сферы деятельности.
Сортируем по алфавиту.
Ограничиваем вывод д 1 строки.
Обрезаем до 3 строчки.

"""

query_5_5 = f'''select
            e.name
            from employers e
            left join employers_industries s on e.id=s.employer_id
            left join industries i on i.id = s.industry_id
            group by e.name
            having count(i.name) = 4
            order by 1 asc
            limit 1
            offset 2

            
            
'''

In [73]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [80]:
# текст запроса

""" _summary_
Выводим кличество компаний (count(e.name)), у которых указано в сфере деятельности 'Разработка программного обеспечения' (i.name like '%Разработка программного обеспечения').
Соединяем таблицы employers, employers_industries и  industries по id работодателя, и по id сферы деятельности.
Группируем по названиям сфер деятельности.

"""

query_5_6 = f'''select
            count(e.name),
            i.name
            from employers e
            left join employers_industries s on e.id=s.employer_id
            left join industries i on i.id = s.industry_id
            where i.name like '%Разработка программного обеспечения'
            group by i.name
            
            

            
            
'''

In [81]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count                                 name
0   3553  Разработка программного обеспечения

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [234]:
import requests 
from bs4 import BeautifulSoup 

url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России' 
response = requests.get(url)

page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер

tabl = page.find("table", class_="standard") #находим таблицу с городами
cities= []
tb = tabl.find("tbody") # в таблице находим ее тело
tr = tb.find_all("tr")  # в находим строки 

for el in tr:
    qqq = list(el.find("a"))  # в каждой строке находим тег "а", преобразуем его в список
    cities.append(qqq[0])       # добавляем список в список городов
cities = tuple(cities[1:])             # находим город и преобразуем сисок в кортеж
print(cities)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [261]:
# текст запроса

""" _summary_
Выводим количество в Регионах вакансий комании Яндекс. А затем объединяем с общим количеством вакансий комании.

"""

query_5_7 = f'''(SELECT 
            a.name AS name,
            COUNT(v.id) AS cnt
            FROM areas a 
            LEFT JOIN vacancies v ON v.area_id = a.id
            LEFT JOIN employers AS e ON e.id = v.employer_id
            WHERE e.name = 'Яндекс'
            AND a.name IN {cities}
            GROUP BY 1
            Order by 2 desc)

            UNION ALL

            (select 'Total', 
            COUNT(v.id)
            FROM areas a 
            LEFT JOIN vacancies v ON v.area_id = a.id
            LEFT JOIN employers AS e ON e.id = v.employer_id
            WHERE e.name = 'Яндекс'
            AND a.name IN {cities}
            GROUP BY 1
            Order by 2 desc)
'''

In [262]:
# результат запроса

df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\2111736488.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt
0             Total  485
1            Москва   54
2   Санкт-Петербург   42
3      Екатеринбург   39
4   Нижний Новгород   36
5       Новосибирск   35
6           Воронеж   32
7         Краснодар   30
8            Самара   26
9               Уфа   26
10           Казань   25
11            Пермь   25
12   Ростов-на-Дону   25
13        Волгоград   24
14       Красноярск   23
15        Челябинск   22
16             Омск   21

***

In [ ]:
# выводы по анализу работодателей




Из проведенного анализа работодателй выяснили:
- Пятерку лидеров, размещающих вакансии.
- На первом месте Яндеск м 485 вакансиями в 181 регионах.
- 410 Работодателей России не имеют вакансий на данный момент, но зарегистрированны в системе.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [104]:
# текст запроса
""" _summary_
Выводим количество вакансий (count(v.name)), которые имеют отношение к данным (ilike (не учитывает регистр) '%data%', %данн%').

"""

query_6_1 = f'''select
            count(v.name)
            from vacancies v
            Where lower(v.name) ilike '%data%' or lower(v.name) ilike '%данн%'
            
            

            
            
'''

In [106]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [126]:
# текст запроса
""" _summary_
Выводим количество вакансий (count(v.name)), которые подходят начинающего дата-сайентиста.

"""

query_6_2 = f'''select
            count(v.name)
            from vacancies v
            Where (lower(v.name) ilike '%data scientist%' 
            or lower(v.name) ilike '%data science%' 
            or lower(v.name) ilike '%исследователь данных%' 
            or lower(v.name) ilike '%machine learning%' 
            or lower(v.name) ilike '%машинн%обучен%'
            or (lower(v.name) ilike '%ML%' and lower(v.name) not ilike '%HTML%'))
            AND (lower(v.name) ilike '%junior%' or v.experience = 'Нет опыта' or employment = 'Стажировка')
            
            

            
            
'''

In [127]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [128]:
# текст запроса
""" _summary_
Выводим количество вакансий (count(v.name)), в которых указаны навыки SQL или postgres.

"""

query_6_3 = f'''select
            count(v.name)
            from vacancies v
            Where (lower(v.name) ilike '%data scientist%' 
            or lower(v.name) ilike '%data science%' 
            or lower(v.name) ilike '%исследователь данных%' 
            or lower(v.name) ilike '%machine learning%' 
            or lower(v.name) ilike '%машинн%обучен%'
            or (lower(v.name) ilike '%ML%' and lower(v.name) not ilike '%HTML%'))
            AND (key_skills ilike '%SQL%' or key_skills ilike '%postgres%')
            
            

            
            
'''

In [129]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [133]:
# текст запроса
""" _summary_
Выводим количество вакансий (count(v.name)), в которых указан навык владения языком Python.

"""

query_6_4 = f'''select
            count(v.name)
            from vacancies v
            Where (lower(v.name) ilike '%data scientist%' 
            or lower(v.name) ilike '%data science%' 
            or lower(v.name) ilike '%исследователь данных%' 
            or lower(v.name) ilike '%machine learning%' 
            or lower(v.name) ilike '%машинн%обучен%'
            or (lower(v.name) ilike '%ML%' and lower(v.name) not ilike '%HTML%'))
            AND (key_skills ilike '%Python%')
            
            

            
            
'''

In [134]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [156]:
# текст запроса
""" _summary_
Считаем среднее количество навыков для вакансий DST.
Для этого из длины строки с навыками длину строки без навыков. 
Исключаем из выборки вакансии без ключевых навыков
"""

query_6_5 = f'''select
            round(avg(length(key_skills) - length(replace(key_skills,CHR(9),''))+1),2)
            from vacancies v
            WHERE(
            (v.name ILIKE '%data scientist%')
            OR (lower(v.name) ILIKE '%data science%')
            OR (v.name ILIKE '%исследователь данных%')
            OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
            OR (v.name ILIKE '%machine learning%')
            OR (v.name ILIKE '%машинн%обучен%')
            AND (key_skills IS NOT NULL)
            )
            
            

            
            
'''

In [157]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [169]:
# текст запроса
""" _summary_

Вычисляем среднюю зп для каждой группы опыта. 

"""

query_6_6 = f'''select
            v.experience,
            ROUND(AVG(coalesce((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)), 0)
            from vacancies v
            WHERE(
            (v.name ILIKE '%data scientist%')
            OR (lower(v.name) ILIKE '%data science%')
            OR (v.name ILIKE '%исследователь данных%')
            OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
            OR (v.name ILIKE '%machine learning%')
            OR (v.name ILIKE '%машинн%обучен%')
            AND (v.salary_from is not null
            or v.salary_to is not null)
            and (v.key_skills IS NOT NULL)
            )
            group by 1
            order by 2 desc
            
            

            
            
'''

In [170]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_19504\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience     round
0         Более 6 лет       NaN
1       От 3 до 6 лет  243115.0
2  От 1 года до 3 лет  145236.0
3           Нет опыта   74643.0

***

In [ ]:
# выводы по предметному анализу

Предметный анализ вакансий показал:
- 1771 вакансий так или иначе относятся к вкансиям с данными.
- 51 вакансии подходят для дата-сайентистов без опыта работы.
- 229 вакансий с ключевыми навыками по sql или postgresql.
- 357 вакансий с навыками владения Python.
- Более 6 навыков указывают работодатели в вакансиях для дата-сайентистов.
- Зарплата для DST варьируется от 74 тыс. руб. для начинающих специалистов, до 243 тыс. руб. для специалистов с опытом от 3 до 6 лет.#

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

In [7]:
""" _summary_
Выводим количество вакансий по самым популярным сферам деятельтности работодателей
"""
 
query_7_1 = f'''select
            i.name,
            count(v.id) as vacancies_cnt
            from industries i
            left join employers_industries ei on i.id = ei.industry_id
            left join vacancies v on v.employer_id = ei.employer_id
            group by i.name
            order by 2 desc
            limit 10
            
            
'''

In [8]:
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_23160\2911634181.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


name  vacancies_cnt
0                Разработка программного обеспечения          12499
1  Системная интеграция,  автоматизации технологи...          11034
2  Интернет-компания (поисковики, платежные систе...           6413
3                                               Банк           2742
4                                    Мобильная связь           1585
5                                 Кадровые агентства           1143
6                              Консалтинговые услуги           1137
7                                 Интернет-провайдер           1095
8                       Розничная сеть (продуктовая)           1034
9                                Фиксированная связь           1028

In [11]:
""" _summary_
Выводим количество вакансий по самым популярным сферам деятельтности работодателей
"""
 
query_7_2 = f'''select
            v.key_skills,
            count(v.id) as vacancies_cnt
            from vacancies v
            where v.key_skills is not null
            group by 1
            order by 2 desc
            limit 10
            
            
'''

In [12]:
df = pd.read_sql_query(query_7_2, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_23160\2993136978.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


key_skills  vacancies_cnt
0                                1С программирование             94
1  Пользователь ПК\tРабота в команде\tРемонт ноут...             83
2  PostgreSQL\tSQL\tLinux\t1С программирование\tM...             71
3          Ответственность\tКлиентоориентированность             48
4  MS PowerPoint\tПодготовка презентаций\tРазрабо...             27
5             1С программирование\t1С: Предприятие 8             26
6  Настройка ПК\tСборка ПК\tРемонт ПК\tНастройка ...             26
7                                    Английский язык             26
8  Грамотная речь\tРабота в команде\tГрамотность\...             25
9                                               Java             24

Проведенный анализ вакансий позволил установить основные треования работодателей к сондидатам.
Мы выяснили самого крупного работодателя - Яндекс.
Установили самый работоспособный регион - Россия.
Провели исследования основных навыков, требуемых работодателей - куда входят не только hard skills, но также  soft skills (работа в команде, ответственность и пр.)
Ваяснили в какой сфере открыто больше всего вакансий.

